File: _projects/mmo/server/data/luascripts/bufs/electricblade_core.lua
local this = BaseClass("ElectricBladeCoreBuf")

--[[
电刀
    3次攻击后，释放一次电击伤害
    +10% 攻速
    +10% 物强
]]--

local scriptName = "电刀_机制"
local triggleConditionTimes = 3

local function reset(self)
    self.idSkillHit = -1
    self.buf = nil
    self.owner = nil

    self.tick = 0
end

local function __ctor(self)
    reset(self)
end

local function onStart(self, buf)
    local owner = buf:Owner()
    local events = owner:GetEvents()

    self.buf = buf
    self.owner = owner
    self.idSkillHit = events:SubscribeNoCheck("onSkillHit", function(skill)
        self:onSkillHit(skill)
    end )
end

local function doEnd(self, buf)
    local owner = buf:Owner()
    local events = owner:GetEvents()

    --print('电刀.onEnd')
    if self.idSkillHit > 0 then
        events:UnsubscribeById("onSkillHit", self.idSkillHit)
        self.idSkillHit = -1
    end
end

local function onEnd(self, buf)
    doEnd(self, buf)
    releaseBufScript(scriptName, self)
end

local function onSkillHit(self, skill)
    --
    if skill:IsBGSkill() then
        --print('bgSkill:'..skill:GetId())
        return
    end
    self.tick = self.tick + 1
    if self.tick >= triggleConditionTimes then
        self.tick = 0

        local owner = self.owner
        --print('电刀伤害')
        owner:CallbackSkill('电刀伤害', 1, owner, owner:GetTarId())
    end
end

this.__ctor = __ctor
this.reset = reset
this.onStart = onStart
this.onEnd = onEnd
this.onSkillHit = onSkillHit

registerBufScript(scriptName, this)
return this

File: _projects/mmo/server/modules/fightscripts/buf/electricblade_core.go
package buf

import (
	"mmo/modules/fight/common"
	"mmo/modules/fight/script"
)

const (
	ebScriptName = "电刀_机制"
	ebHitSkill   = "电刀伤害"
)

func init() {
	mgr.Register(ebScriptName, func() script.IBufScript {
		return &electricbladeScript{}
	})
}

// electricbladeScript
// 电刀
// 3次攻击一次电击伤害
type electricbladeScript struct {
	idSkillHit int
	tick       int
	owner      common.ICharProxy
}

func (s *electricbladeScript) Init(args ...any) {
}

func (s *electricbladeScript) OnStart(proxy script.IBufProxy) {
	owner := proxy.Owner()
	s.owner = owner
	owner.GetEvents().SubscribeWithReceiver("onSkillHit", s, s.onSkillHit)
}

func (s *electricbladeScript) OnEnd(proxy script.IBufProxy) {
	owner := proxy.Owner()
	owner.GetEvents().UnsubscribeWithReceiver("onSkillHit", s, s.onSkillHit)
}

func (s *electricbladeScript) OnTriggle(proxy script.IBufProxy) {
}

func (s *electricbladeScript) onSkillHit(args ...any) {
	if len(args) == 0 {
		return
	}
	skill := args[0].(script.ISkillProxy)
	if skill.IsBGSkill() {
		return
	}
	s.tick++
	if s.tick >= 3 {
		s.tick = 0
		owner := s.owner
		owner.GoCallbackSkill(ebHitSkill, 1, owner, owner.GetTarId())
	}
}
